### ISSUE TITLE
Create a global default text filter to help prevent the use of racist and abusive language

### ISSUE THREAD

**Issue by Servelan** [mastodon/mastodon#31182]
### Pitch

See this post:  https://dair-community.social/@KimCrayton1/112853491446746519.  One of the replies, or replies to replies, was 'consider suicide' from @NateHiggers1488@nationalist.social.   No one should have to stop and consider all the nastiness that might come their way to filter it out manually, and no one should have to deal with crap and then filter when the micro-trauma's already occured.

### Motivation

This feature is needed because:  mods can't catch everything and like police are frequently reactive, not proactive.  Users are subjected to abuse that they have to self-manage when the abuse could have been prevented.  No, it's not 'violating free speech' when you limit the use of the n-word; society, polite or otherwise, including on Mastodon, decides what is acceptable and what is not.  Speech that intends to harm is not acceptable.  

The default filter would be editable by the user, like via checkboxes, so that someone who needs or wants to include this term or that phrase could do so.

The default filter could live with other filters, so the default, plus user defined filters, play nicely together.  If the user filter contradicts the default, the default prevails.

If there's any justice in the world, the same principle would apply to hashtag filters, too, so that a hashtag that's abusive won't go anywhere. #F***N****** should be filterable.

Examples of filter content (obs you want to go for the worst stuff, this is just for rough ideas):
[f*** off & die] (* being a wildcard here)
[n*****] (n-word)
[go k*** yourself]
[consider s******] (as per the aforementioned a**hole)
[tran**]

**Comment by Cassolotl** [comment]
I can think of several non-offensive ways that someone might use any of these words or phrases. Offensiveness is down to context most of the time. These words and phrases aren't inherently and universally unacceptable.

Mastodon's (and the Fediverse's) strength is a much more effective member-to-moderator ratio. We need to rely on that, and trust that people will report because they know that if they do it will work.

Rather than the words/phrases being banned and prevent the words from being posted (which doesn't work anyway because people just bypass filters with asterisks and other junk characters, as the OP did), perhaps the system could just silently put the posts in a list for the moderators' attention. This might be extra work for mods, but perhaps some mods would want to do it?

**Comment by shleeable** [comment]
This is a massive problem and the answer is somewhere between `simple profanity detection` and `contextual awareness LLM`.

https://github.com/brittleru/abusive-language-detection is an example that states it requires 8+GB to work. so that's too much for every instance to run.. but you could have a tool that is a shared service using relays to view the firehose and pick up and report abuse... BUT then you have 7 thousand reports a second made by a bot acting in good intentions...  More noise than signal. 

I'll never expect anybody on my instance to drop the n-word so I could easily add a "global filter" against it.. but it might be worth filtering at the endpoint as well?

So then the question comes to.. instances can have projects, but should clients have adaptable filters to people at risk of abuse? People can opt-into word based abuse filters.

This is one of those things.... wankers need to be responsible for what they send. Moderators have limited resources.. so the people on the receiving end need to be able to detect and block as fast as possible.... I'd imagine something of a reputation system that at risk people can join to collectively block individual accounts on mass. This one would be a community lead effort and would have it's own hardships.

"finding technical problems to social problems" etc etc

**Comment by shleeable** [comment]
![image](https://github.com/user-attachments/assets/e2ae46c5-3c60-4ef0-a988-a5c6a90233dc)

The wanker to target flow.... There are multiple places we can put protection between the wanker and the target.

**Comment by mochsner** [comment]
"""
https://github.com/brittleru/abusive-language-detection is an example that states it requires 8+GB to work. so that's too much for every instance to run.. but you could have a tool that is a shared service using relays to view the firehose and pick up and report abuse... BUT then you have 7 thousand reports a second made by a bot acting in good intentions... More noise than signal.
"""

For the sake of computation and cost effectiveness / scalability for mastodon, I would really love to see something like this as mentioned, where a single service _filters all posts subscribed to by users of instances subscribed to the service_. 
- For a shared centralized mastodon service (or federated network of services, eventually), **pay-per-usage model would be fine**.
  - ...got to start somewhere, and Mastodon is competing with highly centralized big tech companies like Facebook / Twitter, whom are able to move quickly. 

Patreon, etc. would be a great way to try to launch (and maintain) something like this - I am confident it would get a lot of traction on mastodon. Mastodon is at risk of losing a number of marginalized community members, whom I think make the place so great and is what has kept me there. The community is still great, and people *like* what the fediverse offers via Mastodon -- being free of big tech.

Outlining some possibly common sense things, but still an approach to this:

(A) As a base filter would be hateful instances (to prevent needing to look at too many things). For example, anything from e.g. `neonazi-racist.social` should be immediately blocked, once deemed necessary. This is a way for mastodon instances to band together and maintain "shared" blocked instances. Obviously, only extreme offenders should be included here so as to avoid over-censoring. This will be fed into by (B) eventually.
(B) The second filter (and more complex but important one) would be on a per-status basis, using a similar LLM based approach from above to flag things as hate, and store in some combination of database / redis type structure for easy lookup near-future (after enough time, it just recomputes). 
  - Down the road, maintaining an approximate % of hate vs non-hateful posts on a per-instance level could also help feed into what ultimately becomes an (A) blocked instance, to save computation.
  
Just some thoughts. May not be worth anything, and (A) may already be getting worked on elsewhere. But, for the sake of idea sharing and trying to get some traction on this, I wanted to at least jot something down.

**Comment by daprice** [comment]
The proposed feature would also hide posts from the *victims* of this behavior in which they name and discuss the abuse they received. This includes the very same person linked to in the OP.

One of the most common (& valid) complaints I see from posters who are frequent targets of this type of abuse is that the nature of Mastodon/ActivityPub already hides a lot of the abusive comments from bystanders, making it harder for the rest of the community to be aware that it’s happening at all. This filter, applied indiscriminately, would make that problem even worse.

**Comment by shleeable** [comment]
Whenever Mastodon Groups are released in 2077, we can hopefully have dedicated groups for bad actors and abusive users.... more signal

**Comment by mochsner** [comment]
> This filter, applied indiscriminately, would make that problem even worse.

@daprice Sorry if I'm misunderstanding, but if the posts are hidden from the _victim_ (by passing through a filter service before being visible to them, server-side)... how would that make it worse..? Of course, the technical (server load/refactors needed) burden here could be pretty heavy, but from a strictly it being theoretical feasible perspective, I want to make sure things aren't dismissed...

**Comment by daprice** [comment]
@mochsner Sorry if I wasn’t clear, I meant the problem of victims feeling like nobody is aware that abuse is happening, would be worse if they can’t speak openly about it, i.e. because the filter hides *their* posts from other people.

For example if someone makes a post along the lines of “I got called a [n-word] today, that’s awful”, the filter being proposed here would hide that from the majority of other users, because the filter does not and cannot understand context.

Maybe instead of auto-hiding posts, the system could automatically flag posts for moderator review as @Cassolotl suggested earlier.

**Comment by ThisIsMissEm** [comment]
There was also this post from Mekka on keyword filtering: https://hachyderm.io/@mekkaokereke/112897071685037184

i know we've looked into some abusive content filters at IFTAS, but I believe the ones we looked at couldn't differentiate abuse from lived experiences.

**Comment by trwnh** [comment]
I think the issues surrounding false positives can be significantly reduced by having a human reviewer in the loop instead of making it an automatic action. Automated actions in particular have no understanding of context and are vulnerable to the [Scunthorpe problem](https://en.wikipedia.org/wiki/Scunthorpe_problem). Perhaps posts matching the word filter can be surfaced to the moderation team instead of being auto-hidden or auto-deleted?

False negatives are of course much harder to account for, and require contextual awareness to remain ahead of cultural usage. At least at a user level, it should be possible to filter based on keywords such that highly contextual word filters can be engaged at-will for particular users rather than for everyone. Perhaps allowing users to share pre-made filters could help here, to avoid duplicate work? (among other things, #18955 could be related here)

**Comment by FediVideos** [comment]
> I think the issues surrounding false positives can be significantly reduced by having a human reviewer in the loop instead of making it an automatic action

Yes, sounds like a good idea. Admin could have option to make it just flag posts instead of taking action, perhaps varying whether it is automatic or subject to review depending on each filter?

**Comment by ThisIsMissEm** [comment]
> > I think the issues surrounding false positives can be significantly reduced by having a human reviewer in the loop instead of making it an automatic action
> 
> Yes, sounds like a good idea. Admin could have option to make it just flag posts instead of taking action, perhaps varying whether it is automatic or subject to review depending on each filter?

This is also how Bluesky's [indigo automod tool](https://github.com/bluesky-social/indigo/tree/main/automod#readme) works for moderating posts — it just creates a report for them in ozone (which is their moderation tool). It'd probably be possible to use indigo as a starting point for building a mastodon automod tool.

Though it's a lot more advanced, and can take actions based on frequency of occurrences over time and similar stuff

**Comment by FediVideos** [comment]
Another reason why this would be a good fit for Mastodon is there are so many servers and each server has a human moderator. Flagging content could work well because there would always be a human to review it.

This kind of flagging system could actually work a lot better on Mastodon than on commercial social networks, as Mastodon has a much higher ratio of human moderators to users.

**Comment by FediVideos** [comment]
The issue https://github.com/mastodon/mastodon/issues/21306 is perhaps a better fit for those wanting an automatic flagging system?
